### PR TITLE
Publish nodes' out and err as build artefacts

### PR DESCRIPTION
**Pull request by jumaffre**
The CI now publishes the node's logs (`out` and `err`) as a build artefact in Azure Pipelines if a test fails. 

This is quite a crude solution for now as we publish the logs of _all_ the tests in a given job if _any_ test within that job fails (i.e. if the `ctest` task fails). This is probably OK as the logs are only a few KB anyway.

### CODE REVIEW

File: .azure-pipelines-gh-pages.yml [hunk 1]
```diff
@@ -54,4 +54,4 @@ jobs:
     - script: rm deploy_key || true
       displayName: 'Make sure key is removed'
       workingDirectory: sphinx/build/html
-      condition: succeededOrFailed()
+      condition: always()
```
Comment (jumaffre) [inline .azure-pipelines-gh-pages.yml]: This task will get executed even if the job is cancelled.

File: tests/election.py [hunk 2]
```diff
@@ -48,6 +49,8 @@ def run(args):
         network.start_and_join(args)
         current_term = None
 
+        sys.exit(1)
```
Comment (jumaffre) [inline tests/election.py]: Proving that this PR actually works. To revert before merging.

File: .azure-pipelines.yml [hunk 3]
```diff
@@ -47,11 +47,11 @@ resources:
   containers:
     - container: nosgx
       image: ccfciteam/ccf-ci-18.04-oe-0.8.0:latest
-      options: --publish-all --cap-add SYS_PTRACE -v /dev/shm:/ccache
+      options: --publish-all --cap-add SYS_PTRACE -v /dev/shm:/ccache -v /mnt/build:/__w/
```
Comment (jumaffre) [inline .azure-pipelines.yml]: https://github.com/microsoft/azure-pipelines-tasks/issues/12409

**Comment by cimetrics** [comment]
publish_test_workspace@[5218 aka 20200224.38](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5218&view=results) vs master ewma over 30 builds from [4846](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4846&view=results) to [5183](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5183&view=results)
images

File: .azure-pipelines-templates/test.yml [hunk 4]
```diff
@@ -10,6 +10,22 @@ steps:
   displayName: CTest
   workingDirectory: build
 
+# If a test fails, crudely upload nodes' logs for all tests
```
Comment (ashamis) [inline .azure-pipelines-templates/test.yml]: It would also be useful to copy the output of ctest

File: .azure-pipelines-templates/test.yml [hunk 5]
```diff
@@ -10,6 +10,22 @@ steps:
...
+- task: CopyFiles@2
+  inputs:
+    contents: '**/?(out|err)'
```
Comment (ashamis) [inline .azure-pipelines-templates/test.yml]: This is probably outside the scope of this PR but should we produce core dumps on failures and then copy the core files?

File: .azure-pipelines-templates/test.yml (hunk 4 above)
Comment (achamayou) [reply .azure-pipelines-templates/test.yml]: That goes to the UI via publish test results and xslt.

File: .azure-pipelines-templates/test.yml (hunk 5 above)
Comment (achamayou) [reply .azure-pipelines-templates/test.yml]: For the cores to be useful, we’d need to keep the binaries. Which may be fine but is a non trivial amount of storage.

Comment (jumaffre) [reply .azure-pipelines-templates/test.yml]: So I originally tried to do this but this results in quite a lot of storage used (~130 MB per run). See https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=5172&view=artifacts&type=publishedArtifacts

**Comment by jumaffre** [comment]
I also tried to turn on `VERBOSE_LOGGING` for our `Debug` builds on the CI but this results in very verbose unit tests (raft, kv contention). Is there an elegant way to get around this?

**Comment by eddyashton** [comment]
> I also tried to turn on `VERBOSE_LOGGING` for our `Debug` builds on the CI but this results in very verbose unit tests (raft, kv contention). Is there an elegant way to get around this?

We could build with `-DVERBOSE_LOGGING=ON`, but manually set the logging level to `INFO` in those tests? This isn't very elegant.

**Pull request merged**